reactor-core/src/main/java/reactor/core/publisher/SinksSpecs.java

```diff
@@ -164,35 +179,31 @@ public void emitNext(T value) {
 	@Override
 	public final Emission tryEmitNext(T t) {
 		Objects.requireNonNull(t, "t is null in sink.next(t)");
-		if (done) {
-			return Sinks.Emission.FAIL_TERMINATED;
+
+		Thread currentTread = Thread.currentThread();
+		if (WIP.get(this) == 0 && WIP.compareAndSet(this, 0, 1)) {
+			LOCKED_AT.lazySet(this, currentTread);
 		}
-		Thread currentThread = Thread.currentThread();
-		Thread lockedAt = LOCKED_AT.get(this);
-		if (lockedAt != null) {
-			if (lockedAt != currentThread) {
+		else {
+			if (LOCKED_AT.get(this) != currentTread) {
 				return Emission.FAIL_NON_SERIALIZED;
 			}
-		}
-		else if (!LOCKED_AT.compareAndSet(this, null, currentThread)) {
-			return Emission.FAIL_NON_SERIALIZED;
+			WIP.incrementAndGet(this);
 		}
 
-		Emission emission = sink.tryEmitNext(t);
-		LOCKED_AT.compareAndSet(this, currentThread, null);
-		return emission;
+		try {
+			return sink.tryEmitNext(t);
+		}
+		finally {
+			if (WIP.decrementAndGet(this) == 0) {
+				LOCKED_AT.compareAndSet(this, currentTread, null);
+			}
+		}
 	}
 
 	@Override
 	@Nullable
 	public Object scanUnsafe(Attr key) {
-		if (key == Attr.ERROR) {
-			return error;
-		}
-		if (key == Attr.TERMINATED) {
-			return done;
-		}
-
 		return Scannable.from(sink).scanUnsafe(key);
```

Comment:
the `sink` is always `Scannable` so `Scannable.from(sink)` could be turned into `sink` as a polish